adding port v1->v2 TEST-0048 (#13)
ScreaMy7/owasp-mastg: demos/android/MASVS-RESILIENCE/MASTG-DEMO-0x4A/MASTG-DEMO-0x4A.md

Reviewer: The content does not align with the guidelines. Please refer to the guidelines and update them accordingly.https://github.com/OWASP/mastg/blob/master/.github/instructions/mastg-demo.instructions.md

---
platform: android

> title: Uses of Xposed/LSPosed Detection Techniques

Reviewer: ```suggestion
title: Static Detection of Xposed/LSPosed/Frida Hooks using semgrep
```

id: MASTG-DEMO-0x4A
code: [kotlin]
test: MASTG-TEST-0x49
tools: [MASTG-TOOL-0110]
kind: pass
---

## Sample

The snippet below shows sample code that performs four Xposed/LSPosed/Frida detection techniques used by Android apps as anti-instrumentation checks. The checks combine a `PackageManager` lookup for known framework Manager package ids, a reflection-based tripwire on guaranteed-native methods, a `/proc/self/maps` scan for foreign DEX/APK mappings injected into the process, and a stack-trace / thread probe that surfaces framework frames and native worker threads.

> {{ MastgTest.kt # MastgTest_reversed.java # AndroidManifest.xml }}

Reviewer: ```suggestion
The snippet below shows sample code that performs four common Xposed/LSPosed/Frida detection techniques used by Android apps as anti-instrumentation checks: a `PackageManager` presence probe for known Xposed/LSPosed/EdXposed Manager package ids, a `Modifier.isNative` tripwire on guaranteed-native methods, a `/proc/self/maps` scan for foreign DEX/APK mappings injected into the process, and a stack-trace and `/proc/self/task` probe for framework frames and native worker threads.

{{ MastgTest.kt # MastgTest_reversed.java }}
```

## Steps

> Let's run our semgrep rule against the sample code.

Reviewer: ```suggestion
Let's run our @MASTG-TOOL-0110 rule against the sample code.
```

> {{ ../../../../rules/mastg-android-xposed-detection.yaml }}

Reviewer: The correct extension is `.yml` and rename in the `rules` in the folder
```suggestion
{{ ../../../../rules/mastg-android-xposed-detection.yml }}
```

Reviewer: Reference : https://github.com/OWASP/mastg/blob/master/.github/instructions/mastg-rules.instructions.md

{{ run.sh }}

## Observation

> The rule has identified some instances in the code file where detection checks are placed.

Reviewer: ```suggestion
The output contains the locations of all Xposed/LSPosed/Frida detection checks in the code.
```

{{ output.txt }}

## Evaluation

The test case passes because the app statically implements four independent Xposed/LSPosed/Frida detection mechanisms:

- Line 143 declares the list of known Xposed/LSPosed/EdXposed Manager package ids (`de.robv.android.xposed.installer`, `org.lsposed.manager`, `org.meowcat.edxposed.manager`, `io.va.exposed`, `com.solohsu.android.edxp.manager`) consumed by a `PackageManager.getPackageInfo` presence probe.
- Lines 172–174 implement the `Modifier.isNative` tripwire: LSPlant (LSPosed's hooking engine) and Frida's Java bridge clear the `kAccNative` bit when they hook a native method, and this branch flags any audited method whose reflected modifiers no longer report native.
- Line 185 opens `/proc/self/maps` from Java to scan for foreign DEX/APK mappings (e.g. an LSPosed module APK or a Frida agent `.so`) injected into the process address space.

> - Lines 232 and 261 declare the framework/needle string literals (`de.robv.android.xposed`, `org.lsposed.`, `lsphooker_`, `lsplant`, `edxposed`, `re.frida`, `gum-js`, `gmain`, `pool-frida`, `linjector`, …) used to inspect stack traces via `Throwable.getStackTrace` / `Thread.getAllStackTraces`; line 263 enumerates `/proc/self/task` to cross-check `comm` values against those native worker thread needles.

Reviewer: ```suggestion
The test case passes because the app statically implements four independent Xposed/LSPosed/Frida detection mechanisms. Review each of the reported instances:

- Line 143 declares the known Xposed/LSPosed/EdXposed Manager package ids (`de.robv.android.xposed.installer`, `org.meowcat.edxposed.manager`, `io.va.exposed`, `com.solohsu.android.edxp.manager`, `org.lsposed.manager`) consumed by the `PackageManager.getPackageInfo()` presence probe.
- Lines 172-174 implement the `Modifier.isNative` tripwire — branching on a negated `Modifier.isNative()` check against a reflected method's modifiers detects LSPlant (LSPosed) and Frida's Java bridge clearing `kAccNative` when they hook a native method.
- Line 185 opens `/proc/self/maps` from Java to scan for a foreign module APK (e.g., an LSPosed module) or a Frida agent `.so` mmapped into the process.
- Line 232 declares the instrumentation-framework needle list (`de.robv.android.xposed`, `org.lsposed.lspd`, `org.lsposed.`, `lsphooker_`, `lsplant`, `edxposed`, `re.frida`, `gum-js`) used to inspect Java stack traces from `Thread.getAllStackTraces()` and from forced exceptions on `PackageManager`, `Runtime.exec`, and `File.exists`.
- Line 261 declares the native-worker-thread needle list (`gum-js`, `gmain`, `pool-frida`, `frida`, `linjector`, `lspd`, `xposed`, `lsphooker`) matched against each thread name read from `/proc/self/task/<tid>/comm`.
- Line 263 enumerates `/proc/self/task` to feed those `comm` reads, surfacing Frida's Gum/JS worker threads and LSPosed daemon threads injected into the process.
```

Author: The demo file is updated

{{ MastgTest.kt # MastgTest_reversed.java # AndroidManifest.xml }}

## Steps

Run the static analysis rule against the decompiled code.

{{ ../../../../rules/mastg-android-xposed-detection.yaml }}

{{ run.sh }}

## Observation

The output shows the semgrep rule matching all four detection routines in the reversed Java.

{{ output.txt }}

## Evaluation

The test passes because the app statically implements four independent Xposed/LSPosed/Frida detection mechanisms — semgrep flags:

- `PackageManager.getPackageInfo` against the known Manager package ids (`de.robv.android.xposed.installer`, `org.lsposed.manager`, `org.meowcat.edxposed.manager`, `io.va.exposed`, `com.solohsu.android.edxp.manager`).
- The `Modifier.isNative` tripwire on the audited guaranteed-native methods.
- The `/proc/self/maps` reader paired with `/data/app/` inspection.
- The stack-trace / `/proc/self/task` probe via `Throwable.getStackTrace` / `Thread.getAllStackTraces`.

!!! note

>    This is a "kind: pass" demo for static presence only: the sample contains the detection mechanisms expected by the test. The dynamic counterpart @MASTG-DEMO-0x4B covers a Frida-based runtime bypass that defeats all four of these checks.

Reviewer: Remove
```suggestion

```
